DSD-418: Chakra/DS Button component

EdwinGuzman opened the pull request:
Fixes JIRA ticket [DSD-418](https://jira.nypl.org/browse/DSD-418)

## This PR does the following:
- Uses Chakra's `Button` component to compose the DS `Button` component.
- This is a WIP but setting the PR to view the preview Storybook instance.

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have updated the Storybook documentation accordingly.
- [x] All new and existing tests passed.

### Front End Review:

<!--- This step is done AFTER creating a PR. -->
<!--- Netlify creates a static Storybook preview URL once the PR is created. -->
<!--- Copy the URL to the relevant Storybook page here. -->

- [x] View [the example in Storybook]()

EdwinGuzman committed:
Using Chakra's Button component to compose a DS Button component

EdwinGuzman commented:
Update from meeting:
- we'll remove SCSS and move forward with css-in-js
- there's going to be a variant for the red donate button style
- We need to work on the global styles and themes for spacing and other rules ASAP
- I'm going to stick with `buttonType` over `variant` for the prop type (for now)
- for a separate PR we'll document how to write style props and what props are available for DS component and why there's a need to restrict adding style props to Chakra components that are re-exported by DS

I'll continue working on this PR based on the above.

@wluisi @hokei @bigfishdesign13

hokei commented:
> there's going to be a variant for the red donate button style

Does that mean we're purposely not exposing the other props, or does that mean we're still going to expose them but make it clear in the docs that those other props should be used as a last resort?  @bigfishdesign13 

(PS we don't have to make clear what the chakra props are in the stories, we can just link to Chakra and be like, "these are available but use them as a last resort")

EdwinGuzman commented:
> Does that mean we're purposely not exposing the other props, or does that mean we're still going to expose them but make it clear in the docs that those other props should be used as a last resort? @bigfishdesign13

For now, none of the style props are actually available for the DS `Button` component since that's our custom component. In this case, we're not exposing the other props but it's not intentional. We just haven't fully decided on whether we should use a `{...rest}` approach for any and all other props or keep using the attribute. Maybe we can hold off on that decision until more components are built and we see, through implementation, what the best approach is.

> (PS we don't have to make clear what the chakra props are in the stories, we can just link to Chakra and be like, "these are available but use them as a last resort")

I'm already doing this (except for the "use them as a last resort" message) in the other PR the re-exports Chakra components directly from the DS - https://pr640-1cyzfp5o9mouqwm0dsk0lxabk2hgoxgl.tugboat.qa/?path=/docs/chakra-layout-box--box
There's a link to the style props documentation in the Story above.

EdwinGuzman committed:
Second iteration of the Button w Chakra

EdwinGuzman committed:
Merging and fixing conflicts from DSD-414/chakra-unit

EdwinGuzman commented:
This is no longer a WIP and is ready for a full review.

EdwinGuzman commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr642-0kvfxcldirptzwjy7ykojz9esatodof9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/611d668c331d463cd063d4de

EdwinGuzman committed:
Extending Chakra's base theme with NYPL's theme.

EdwinGuzman committed:
Reverting back to prev commit

EdwinGuzman commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr642-0kvfxcldirptzwjy7ykojz9esatodof9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/611d668c331d463cd063d4de

EdwinGuzman committed:
Extending Chakra's base theme with NYPL's theme.

EdwinGuzman committed:
Initial implementation of Chakra's Tab components

EdwinGuzman committed:
Fixing tests

EdwinGuzman committed:
Updating colors obj

hokei reviewed (approved):
This looks good to me!  Most of my questions are around CSS going forward, which seems out of scope for this PR.

nypl-wluisi commented on src/components/Button/Button.tsx:
```diff
@@ -16,84 +22,82 @@ interface ButtonProps {
+  return (
+    <ChakraButton
+      id={id}
+      className={bem(baseClass, [], "", [className])}
```
Are we keeping `className` and bem related around temporarily for backwards compatibility reasons?

EdwinGuzman replied:
Yes, and also because we haven't completely moved away from bem so maybe this should stick around until we no longer need it.

nypl-wluisi replied:
Should we consider adding a deprecation flag comment for this and other parts that we're planning on getting rid of? Something like:

```js
// @deprecated since version 2.3.
// Will be deleted in version 3.0.
 ```

bigfishdesign13 replied:
@nypl-wluisi Doing this on a code level seems untenable.  If we are going to flag things like this, what about adding a `Deprecated` page to the `Documentation` section in Storybook?

nypl-wluisi replied:
Yeah, that's a good point. We can discuss more, but we need a list somewhere of all props and other things that we plan on no longer using with the switch to Chakra. And then we also need to decide if something like `bem` and `className` are things we are going to move away from, which I think we are?

EdwinGuzman replied:
Yes, I think we should move away from `bem` with this Chakra update but it'll be a slow transition. I do like that the current function with `className={bem(baseClass, [], "", [className])}` (where classname = "newclass") will render `"button newclass". We might have to replace this with 
```
className={`${baseClass} ${className}`}
```

nypl-wluisi commented:
This might be pre-existing, but the naming of the props `buttonType` and `type` is a bit confusing. `buttonType` prop controls visual changes for the component itself, where as `type` sets the actual `type` attribute on the button HTML element?

nypl-wluisi commented:
Search Bar component has some styling issues, related to this button change: https://pr642-0kvfxcldirptzwjy7ykojz9esatodof9.tugboat.qa/?path=/story/components-form-elements-searchbar--basic

<img width="1366" alt="search-bar-bug" src="https://user-images.githubusercontent.com/62392174/130671963-e4f97e7d-66fb-4ced-8658-fda81fe51bbd.png">

hokei commented:
> Search Bar component has some styling issues, related to this button change: 

This is the kind of thing I'd love to figure out if we can catch using automated tests - would a snapshot catch this?

bigfishdesign13 commented on src/components/Button/_Button.scss:
```diff
@@ -1,48 +1,12 @@
   display: flex;
   justify-content: center;
-  padding: var(--space-xs) var(--space-s);
+  padding: var(--space-xs) var(--space-s) !important;
```
What is necessitating the `!important`?

EdwinGuzman replied:
This is only really being used in the `Accordion` component to style the input. Some of Chakra's inline css rules take precedence so it needs the `!important`. But once we update the `Accordion` component, this can be removed.

bigfishdesign13 commented on src/theme/components/button.ts:
```diff
@@ -0,0 +1,91 @@
+    cursor: "pointer",
+    color: "ui.white",
+    justifyContent: "center",
+    padding: "var(--space-xs) var(--space-s)",
```
This is something that is dependent on another PR, but should you replace the `space` CSS vars with the Chakra data objects before merging this?

EdwinGuzman replied:
This is fixed in #648 so I'll leave it there to see the progression of the Chakra integration for this component.

EdwinGuzman committed:
Init random thoughts because busy today

EdwinGuzman committed:
Merge branch 'DSD-414/chakra-init' into DSD-418/chakra-button

* DSD-414/chakra-init: (35 commits)
  updates based on review comments
  Explicitly adding postcss 8
  Explicitly adding postcss 8
  reworked image handling and other corrections
  Fixing tests
  fixed based on comments
  enhanced Form story to better show spacing variants
  Fixing rendering bug in the Accordion story
  Adding the optional 'showOptReqLabel' prop in the Select component to optionally render Required/Optional in its label
  updated tests and CHANGELOG
  custom form spacing
  Better order of props in the Label component
  Using and passing a prop to conditionally render the text 'Required' or 'Optional' in the , , and  components.
  adjusted image sizing
  added card image sizes
  updated stories
  changelog and index exports
  added imageFlip and center props
  cleaned ArgsTable and added more props
  added interfaces and type casting
  ...

EdwinGuzman committed:
Merging and fixing conflicts from development

EdwinGuzman commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr642-0kvfxcldirptzwjy7ykojz9esatodof9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/611d668c331d463cd063d4de

bigfishdesign13 reviewed (commented):
I think this looks good as we start shifting away from the SCSS.  I just have some minor questions.

EdwinGuzman commented:
> This might be pre-existing, but the naming of the props buttonType and type is a bit confusing. buttonType prop controls visual changes for the component itself, where as type sets the actual type attribute on the button HTML element?

Yeah, it's confusing but you're right about how it works.

EdwinGuzman commented:
> This is the kind of thing I'd love to figure out if we can catch using automated tests - would a snapshot catch this?

Yes, exactly. We just now need to implement more snapshot tests in order to catch these issues.

EdwinGuzman committed:
Adding documentation for Chakra and its use with DS components. Init draft

EdwinGuzman committed:
Fixing test

EdwinGuzman commented:
Snapshot testing catching visual differences:

I reverted the Button component, added a snapshot test for the `SearchBar` component, un-reverted the Button changes, and then ran the tests again and it did fail:

<img width="581" alt="Screen Shot 2021-08-25 at 3 40 27 PM" src="https://user-images.githubusercontent.com/1280564/130854776-b93da9ec-bb55-435d-be0b-318b3ac0c2c6.png">

Ignore the second error with the div/button -- this one fails because the `id` is autogenerated and it will always be different and we need to fix this separately. But you can see that the `button` component has new classes so we can infer that maybe it also changed visually. This is helpful but we need to set time aside first to write snapshot tests in order to catch more of these issues in the future. I need to revert this snapshot test since the `Icon` component is its own issue, but I'll fix the button so it looks right in the `SearchBar` component.

EdwinGuzman committed:
Fixing stories that use the button and updating button styles

EdwinGuzman commented:
Updated.

EdwinGuzman committed:
Fixing global Chakra default svg style rule.

EdwinGuzman commented:
@wluisi  the latest commit should fix the svg issue in the header nav buttons, but it's global so it should fix it everywhere else.

EdwinGuzman committed:
Updating table of contents

EdwinGuzman committed:
Adding more comments to make it clear not to use the Chakra color shades

EdwinGuzman committed:
Merging and fixing conflicts from DSD-414/chakra-init

EdwinGuzman committed:
Merging and fixing conflicts from DSD-418/chakra-button

EdwinGuzman committed:
Merging and fixing conflicts from DSD-417/chakra-custom-theme

EdwinGuzman committed:
Merging and fixing conflicts from DSD-29/tabs-init

EdwinGuzman committed:
Fixing documentation based on feedback

EdwinGuzman committed:
Fixing date test

EdwinGuzman committed:
Fixing Tabs so that the data array can take HTML in the content attribute as a string or straight HTML

EdwinGuzman committed:
Merge pull request #654 from NYPL/DSD-419/chakra-documentation

DSD-419: Chakra Documentation

EdwinGuzman committed:
Merge pull request #652 from NYPL/DSD-29/tabs-init

DSD-29: Tab components

EdwinGuzman committed:
Merge pull request #648 from NYPL/DSD-417/chakra-custom-theme

DSD-417: Extending Chakra's base theme with NYPL's theme.
